code-dot-org/code-dot-org
# [JIMT][SL-592] tutorial explorer dialog focus

## Summary

In the list of tutorials, when you pop open the tutorial detail dialog you would not be able to tab between items on the dialog, making it less accessible.

For starters, _technically_ you could tab on the dialog - the issue was that the dialog actually lived "early" in the page, so if you shift-tabbed a number of times until you fell off the start of the tutorial list, you'd then tab your way back up through the items on the modal until you eventually tabbed your way through it and fell off the other end.

This mod does a few things:
- The dialog itself is wrapped in `focus-trap-react`, which'll _very_ greedily grab onto focus and force it into the dialog, starting on the `X` button in the upper right.
- If the user has focused on a part of the dialog and hits the left/right arrow to go to the next tutorial, the focus will remain on whatever they had previously focused. If the focused item disappears (such as with "Teacher notes"), the focus will hang around in that area and the next tab will take it to the next available item.
- If the user clicked on a tutorial and then close the detail, the tutorial they had clicked on would gain focus. However, if the user changed tutorial via left/right arrow, focus would remain on the previously selected tutorial. Now when the detail is closed, the currently selected tutorial is focused and scrolled into view if necessary. It won't focus or scroll the selected tutorial until the dialog closes.

A few other points raised from this work:
- In some poking around, it looks like most of our dialogs don't properly behave in this manner and tabbing in them will lose focus, which is not ADA compliant. `focus-trap-react` is a wrapper around `focus-trap`, which we can use to add in this ability to our non-react based dialogs. This should give us a consistent experience until we get it refactored out.
- IMHO, the explore page would benefit from having the list of filters be static while the tutorials themselves scroll. As is, the user needs to scroll all the way back to the top.
- The page could also probably benefit from a refactor - it's just older style class based components, and more modern react might be easier to manage. I ended up deciding it wasn't urgent enough to push through atm.
- If you close the tutorial detail view by clicking outside of the modal, your focus will _not_ update, and in fact you will lose focus entirely.

## Links

jira ticket: [SL-592](https://codedotorg.atlassian.net/jira/software/c/projects/SL/issues/SL-592)

## Testing story

Basically, check all the scenarios listed above.
- Open tutorial detail, confirm you can tab through links w/o scrolling the page.
- Switch tutorials and confirm focus moves to the selected tutorial upon close.



## Deployment strategy

## Follow-up work



## Privacy



## Security



## Caching

## PR Checklist:
